src/essence/essence.js

```diff
@@ -125,6 +126,147 @@ $(document.body).keydown(function (e) {
 var essence = {
     configData: null,
     hasSwapped: false,
+    ws: null,
+    initialWebSocketRetryInterval: 60000, // 1 minute
+    webSocketRetryInterval: 60000, // Start with this time and double if disconnected
+    webSocketPingInterval: null,
+    connectWebSocket: function(path, initial) {
+        // https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/readyState
+        if (essence.ws === undefined || essence.ws === null
+                || (essence.ws && essence.ws.readyState === 3)) {
+
+            essence.initWebSocket(path)
+
+            // If we're trying to start the WebSocket for the first time, we know we're not connected already
+            // so  we don't need to retry to connect yet
+            if (!initial) {
+                clearInterval(essence.webSocketPingInterval)
+                essence.webSocketRetryInterval = essence.webSocketRetryInterval * 2
+                essence.webSocketPingInterval = setInterval(essence.connectWebSocket, essence.webSocketRetryInterval, path, false) // 10 seconds
+            }
+        }
+    },
+    initWebSocket: function(path) {
+        essence.ws = new WebSocket(path)
+
+        essence.ws.onerror = function(e) {
+            console.log(`Unable to connect to WebSocket at ${path}`)
+
+            M.Toast.dismissAll()
+
+            M.toast({
+                html: `Not connected to WebSocket. Will retry in ${(essence.webSocketRetryInterval / 60000).toFixed(2)} minutes...`,
+                displayLength: 10000,
+                classes: "mmgisToast failure",
+            });
+        }
+
+        essence.ws.onopen = function () {
+            console.log('Websocket connection opened...')
+
+            UserInterface_.removeLayerUpdateButton()
+
+            M.Toast.dismissAll()
+
+            if (essence.webSocketRetryInterval > essence.initialWebSocketRetryInterval) {
+                M.toast({
+                    html: "Successfully connected to WebSocket",
+                    displayLength: 160000000,

```

Comment:
Nice! And this is a lot of milliseconds.

Reply:
Crud, I would say something about Monday but it's Tuesday now! Fixed with 3b4cdd011e0296db05cdc2d80ba92e358eeb2164!